python/amici/sbml_import.py

```diff
@@ -1501,7 +1499,105 @@ def _add_conservation_for_non_constant_species(
+
+        try:
+            S = np.asarray(self.stoichiometric_matrix, dtype=float)
+        except TypeError:
+            # Due to the numerical algorithm currently used to identify
+            #  conserved quantities, we can't have symbols in the
+            #  stoichiometric matrix
+            warnings.warn("Conservation laws for non-constant species in "
+                          "combination with parameterized stoichiometric "
+                          "coefficients are not currently supported "
+                          "and will be turned off.")
+            return []
+
+        if any(rule.getTypeCode() == sbml.SBML_RATE_RULE
+               for rule in self.sbml.getListOfRules()):
+            # see SBML semantic test suite, case 33 for an example
+            warnings.warn("Conservation laws for non-constant species in "
+                          "models with RateRules are not currently supported "
+                          "and will be turned off.")
+            return []
+
+        # Determine rank via SVD
+        rank = matrix_rank(S) if S.shape[0] else 0
+        if rank == S.shape[0]:
+            return []
+        kernel = nullspace_by_rref(S.T)
+        # Check dimensions - due to numerical errors, nullspace_by_rref may
+        #  fail in certain situations
+        if kernel.shape[0] != S.shape[0] - rank:
+            raise AssertionError(
+                "Failed to determine all conserved quantities "
+                f"(found {kernel.shape[0]}, expected {S.shape[0] - rank}). "
+                "Try another algorithm, disable detection of conservation "
+                "laws, or submit a bug report along with the model."
+            )
+        kernel = rref(kernel)
+        raw_cls = []
+        for row in kernel:
+            state_idxs = [i for i, coeff in enumerate(row) if coeff]
+            coefficients = [coeff for coeff in row if coeff]
+            print(state_idxs, coefficients)
+            raw_cls.append((state_idxs[0], state_idxs, coefficients),)
+
+        return raw_cls
+
+    def _add_conservation_for_non_constant_species(
+        self,
+        ode_model: ODEModel,
+        conservation_laws: List[ConservationLaw]
+    ) -> List[int]:
+        """Add non-constant species to conservation laws
+
+        :param ode_model:
+            ODEModel object with basic definitions
+        :param conservation_laws:
+            List of already known conservation laws
+        :returns:
+            List of species indices which later remain in the ODE solver
+        """
+        # indices of retained species
+        species_solver = list(range(ode_model.num_states_rdata()))
+
+        algorithm = os.environ.get("AMICI_EXPERIMENTAL_SBML_NONCONST_CLS", "")
+        if algorithm.lower() == "demartino":
+            raw_cls = self._get_conservation_laws_demartino(ode_model)
+        else:
+            raw_cls = self._get_conservation_laws_new()
```

Comment:
Ideas for better algorithm names? 

I'd still keep the previous algorithm as an option, at least for another while.